sw/device/lib/testing/usb_logging.c

```diff
@@ -0,0 +1,550 @@
+  ctx->free = buf;
+}
+
+static status_t buffer_send(usb_logging_ctx_t *ctx, usb_logging_stream_t *s) {
+  TRY_CHECK(ctx && s && s->send_buf);
+
+  const uint8_t *data = s->send_buf->data;
+  uint16_t len = s->send_buf->bytes_used;
+
+  if (kDevLogging) {
+    LOG_INFO("sending %p %p %u\n", ctx, data, len);
+  }
+
+  TRY(usb_testutils_transfer_send(ctx->usbutils, s->ep, data, len,
+                                  kUsbTestutilsXfrDoubleBuffered));
+  // Have at least started to send the buffer...
+  s->sending = true;
+  return OK_STATUS();
+}
+
+static status_t buffer_completed(usb_logging_ctx_t *ctx,
+                                 usb_logging_stream_t *s) {
+  TRY_CHECK(s->wr_buf);
+
+  // Make the buffer available for sending.
+  usb_logging_buf_t **prev = &s->send_buf;
+  while (*prev) {
+    prev = &(*prev)->next;
+  }
+  *prev = s->wr_buf;
+  s->wr_buf = NULL;
+
+  // Attempt to send.
+  status_t res = OK_STATUS();
+  if (!s->sending) {
+    res = buffer_send(ctx, s);
+  }
+  return res;
+}
+
+// Indicates whether a stream is still busy, when attempting to finalize.
+static inline bool stream_busy(usb_logging_ctx_t *ctx,
+                               usb_logging_stream_t *s) {
+  if (s->wr_buf) {
+    buffer_completed(ctx, s);
+  }
+  return (s->send_buf != NULL);
+}
+
+// A buffer of logging data has been received by the host; we may try to
+// send another buffer if one is available.
+static status_t tx_done(void *s_v, usb_testutils_xfr_result_t result) {
+  usb_logging_stream_t *s = (usb_logging_stream_t *)s_v;
+  usb_logging_ctx_t *ctx = s->ctx;
+
+  s->sending = false;
+
+  // Return this buffer to the free list.
+  usb_logging_buf_t *buf = s->send_buf;
+  TRY_CHECK(buf != NULL);
+  s->send_buf = buf->next;
+  buffer_release(ctx, buf);
+
+  // Send the next, if there is one
+  if (s->send_buf) {
+    TRY(buffer_send(ctx, s));
+  }
+  return OK_STATUS();
+}
+
+// Callback function from the usb_testutils layer; presently this is called
+// every 16ms, provided that usb_testutils_poll is being called often.
+static status_t tx_flush(void *s_v) {
+  usb_logging_stream_t *s = (usb_logging_stream_t *)s_v;
+  usb_logging_ctx_t *ctx = s->ctx;
+
+  // We use this to prevent incomplete buffers of log data sitting here
+  // indefinitely and not reaching the host.
+  usb_logging_buf_t *buf = s->wr_buf;
+  if (buf && ibex_timeout_check(&s->flush_time)) {
+    if (buf->bytes_used > 0u) {
+      // Send everything that we have thus far collected in this buffer,
+      buffer_completed(ctx, s);
+    }
+    // Earliest time at which we shall again attempt flushing of log data on
+    // this stream.
+    s->flush_time = ibex_timeout_init(kUSBLogFlushInterval);
+  }
+
+  return OK_STATUS();
+}
+
+// Internal function for all USB logging.
+static status_t usb_logging_send(usb_logging_ctx_t *ctx,
+                                 usb_logging_stream_t *s, const uint8_t *data,
+                                 size_t len) {
+  size_t idx = 0U;
```

Comment:
Instead of using idx and repeat arithmetic below, it might make more sense to increment `data` ptr and shrink `len`?

Reply:
Ok, adjusted now that other people are comfortable with that way that I would naturally have expressed it myself, modifying function arguments :)